src/lib/libc/stdio/fgetws.3
.\"	$OpenBSD: fgetws.3,v 1.2 2005/06/17 22:03:18 jmc Exp $
.\"
.\" $NetBSD: fgetws.3,v 1.2 2003/08/07 16:43:23 agc Exp $
.\"
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Chris Torek and the American National Standards Committee X3,
.\" on Information Processing Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)fgets.3	8.1 (Berkeley) 6/4/93
.\"
.\" Original version ID:
.\" FreeBSD: src/lib/libc/stdio/fgets.3,v 1.16 2002/05/31 05:01:17 archie Exp
.\" FreeBSD: src/lib/libc/stdio/fgetws.3,v 1.2 2002/09/06 11:23:55 tjr Exp
.\"
.Dd August 6, 2002
.Dt FGETWS 3
.Os
.Sh NAME
.Nm fgetws
.Nd get a line of wide characters from a stream
.Sh SYNOPSIS
.In stdio.h
.In wchar.h
.Ft "wchar_t *"
.Fn fgetws "wchar_t *ws" "int n" "FILE *fp"
.Sh DESCRIPTION
The
.Fn fgetws
function
reads at most one less than the number of characters specified by
.Fa n
from the given
.Fa fp
and stores them in the wide character string
.Fa ws .
Reading stops when a newline character is found,
at end-of-file, or error.
The newline, if any, is retained.
If any characters are read and there is no error, a
.Ql \e0
character is appended to end the string.
.Sh RETURN VALUES
Upon successful completion,
.Fn fgetws
returns
.Fa ws .
If end-of-file occurs before any characters are read,
.Fn fgetws
returns
.Dv NULL
and the buffer contents remain unchanged.
If an error occurs,
.Fn fgetws
returns
.Dv NULL
and the buffer contents are indeterminate.
The
.Fn fgetws
function
does not distinguish between end-of-file and error, and callers must use
.Xr feof 3
and
.Xr ferror 3
to determine which occurred.
.Sh ERRORS
.Bl -tag -width Er
.It Bq Er EBADF
The given
.Fa fp
argument is not a readable stream.
.It Bq Er EILSEQ
The data obtained from the input stream does not form a valid
multibyte character.
.El
.Pp
The function
.Fn fgetws
may also fail and set
.Va errno
for any of the errors specified for the routines
.Xr fflush 3 ,
.Xr fstat 2 ,
.Xr read 2 ,
or
.Xr malloc 3 .
.Sh SEE ALSO
.Xr feof 3 ,
.Xr ferror 3 ,
.Xr fgets 3
.Sh STANDARDS
The
.Fn fgetws
function
conforms to
.St -p1003.1-2001 .


src/lib/libc/stdio/fgetws.c
/*	$OpenBSD: fgetws.c,v 1.2 2005/08/08 08:05:36 espie Exp $	*/
/* $NetBSD: fgetws.c,v 1.1 2003/03/07 07:11:37 tshiozak Exp $ */

/*-
 * Copyright (c) 2002 Tim J. Robbins.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * Original version ID:
 * FreeBSD: src/lib/libc/stdio/fgetws.c,v 1.4 2002/09/20 13:25:40 tjr Exp
 *
 */

#include <errno.h>
#include <stdio.h>
#include <wchar.h>
#include "local.h"

__RCSID("$MirOS$");

wint_t __fgetwc_unlock(FILE *);

wchar_t *
fgetws(wchar_t *ws, int n, FILE *fp)
{
	wchar_t *wsp;
	wint_t wc;

	flockfile(fp);
	_SET_ORIENTATION(fp, 1);

	if (n <= 0) {
		errno = EINVAL;
		goto error;
	}

	wsp = ws;
	while (n-- > 1) {
		if ((wc = __fgetwc_unlock(fp)) == WEOF && errno == EILSEQ) {
			goto error;
		}
		if (wc == WEOF) {
			if (wsp == ws) {
				/* EOF/error, no characters read yet. */
				goto error;
			}
			break;
		}
		*wsp++ = (wchar_t)wc;
		if (wc == L'\n') {
			break;
		}
	}

	*wsp++ = L'\0';
	funlockfile(fp);

	return (ws);

error:
	funlockfile(fp);
	return (NULL);
}
